File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/fragment/home/HomeViewModel.java
package com.zuk0.gaijinsmash.riderz.ui.fragment.home;

import android.app.Application;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.room.Room;
import android.content.Context;
import android.os.CountDownTimer;
import android.os.Parcelable;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.zuk0.gaijinsmash.riderz.R;
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite;
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.Bsa;
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.BsaXmlResponse;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Etd;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.EtdXmlResponse;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Trip;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.TripJsonResponse;
import com.zuk0.gaijinsmash.riderz.data.local.room.database.FavoriteDatabase;
import com.zuk0.gaijinsmash.riderz.data.remote.repository.BsaRepository;
import com.zuk0.gaijinsmash.riderz.data.remote.repository.EtdRepository;
import com.zuk0.gaijinsmash.riderz.data.remote.repository.TripRepository;
import com.zuk0.gaijinsmash.riderz.utils.SharedPreferencesUtils;
import com.zuk0.gaijinsmash.riderz.utils.StationUtils;
import com.zuk0.gaijinsmash.riderz.utils.TimeDateUtils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class HomeViewModel extends ViewModel {

    private Application mApplication;
    private TripRepository mTripRepository;
    private BsaRepository mBsaRepository;
    private EtdRepository mEtdRepository;
    private List<Bsa> mBsaList;
    private Parcelable mEtdState;

    @Inject
    public HomeViewModel(Application application,
                         TripRepository tripRepository,
                         BsaRepository bsaRepository,
                         EtdRepository etdRepository) {
        mApplication = application;
        mTripRepository = tripRepository;
        mBsaRepository = bsaRepository;
        mEtdRepository = etdRepository;
    }

    public void setEtdState(Parcelable state) {
        mEtdState = state;
    }

    public Parcelable getState() {
        return mEtdState;
    }

    public void setBsaList(List<Bsa> bsaList) {
        mBsaList = bsaList;
    }

    public List<Bsa> getBsaList() {
        return mBsaList;
    }

    public LiveData<BsaXmlResponse> getBsaLiveData() {
        return mBsaRepository.getBsa();
    }

    public int getHour() {
        return TimeDateUtils.getCurrentHour();
    }

    public boolean is24HrTimeOn(Context context) {
        return SharedPreferencesUtils.getTimePreference(context);
    }

    private String initTime(boolean is24HrTimeOn, String time) {
        String result;
        if(is24HrTimeOn) {
            result = TimeDateUtils.format24hrTime(time);
        } else {
            result = TimeDateUtils.convertTo12Hr(time);
        }
        return result;
    }

    String initMessage(Context context, boolean is24HrTimeOn, String time) {
        return context.getResources().getString(R.string.last_update) + " " + initTime(is24HrTimeOn, time);
    }

    void initPic(Context context, int hour, ImageView imageView) {
        if (hour < 6 || hour >= 21) {
            // show night picture
            Glide.with(context)
                    .load(R.drawable.sf_night)
                    .into(imageView);
        } else if (hour >= 17) {
            // show dusk picture
            Glide.with(context)
                    .load(R.drawable.sf_dusk)
                    .into(imageView);
        } else {
            Glide.with(context)
                    .load(R.drawable.sf_day)
                    .into(imageView);
        }
    }

    boolean doesPriorityExist() {
        FavoriteDatabase db = Room.databaseBuilder(mApplication, FavoriteDatabase.class,
                "favorites").allowMainThreadQueries().build();
        int count = db.getFavoriteDAO().getPriorityCount();
        db.close();
        return count > 0;
    }

    public Favorite getFavorite() {
        FavoriteDatabase db = Room.databaseBuilder(mApplication, FavoriteDatabase.class,
                "favorites").allowMainThreadQueries().build();
        Favorite favorite = db.getFavoriteDAO().getPriorityFavorite();
        db.close();
        return favorite;
    }

    LiveData<EtdXmlResponse> getEtdLiveData(String origin) {
        String originAbbr = StationUtils.getAbbrFromStationName(origin);
        return mEtdRepository.getEtd(originAbbr);
    }

    public Favorite createFavoriteInverse(List<Trip> trips, Favorite favorite) {
        ArrayList<String> trainHeaders = new ArrayList<>();
        for(Trip trip : trips) {
            String header = trip.getLegList().get(0).getTrainHeadStation();
            if(!trainHeaders.contains(header)) {
                trainHeaders.add(header);
                Log.i("HEADER", header);
            }
        }
        Favorite favoriteInverse = new Favorite();
        favoriteInverse.setOrigin(favorite.getDestination());
        favoriteInverse.setDestination(favorite.getOrigin());
        favoriteInverse.setTrainHeaderStations(trainHeaders);
        return favoriteInverse;
    }

    LiveData<TripJsonResponse> getTripLiveData(String origin, String destination) {
        return mTripRepository.getTrip(StationUtils.getAbbrFromStationName(origin),
                StationUtils.getAbbrFromStationName(destination),"TODAY" ,"NOW");
    }

    List<Estimate> getEstimatesFromEtd(Favorite favorite, List<Etd> etds) {
        List<Estimate> results = new ArrayList<>();
        for(Etd etd : etds) {
            if(favorite.getTrainHeaderStations().contains(etd.getDestinationAbbr())) {
                Estimate estimate = etd.getEstimateList().get(0);
                estimate.setOrigin(favorite.getOrigin());
                estimate.setDestination(favorite.getDestination());
                results.add(estimate);
            }
        }
        return results;
    }

    public static CountDownTimer beginTimer(TextView textView, int minutesLeft) {
        long untilFinished = (long) (minutesLeft * 60000);
        return new CountDownTimer(untilFinished, 1000) {
            String minutes = textView.getContext().getResources().getString(R.string.minutes);
            //String seconds = textView.getContext().getResources().getString(R.string.seconds);

            @Override
            public void onTick(long millisUntilFinished) {
                String remainingTime = String.valueOf(millisUntilFinished / 60000)+ " " + minutes;
                textView.setText(remainingTime);
            }

            @Override
            public void onFinish() {
                textView.setText(textView.getContext().getResources().getString(R.string.leaving));
                // todo: if leaving, wait 1 minute, then destroy itemListRow.
            }
        }.start();
    }
}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/fragment/home/HomeViewModel.kt
package com.zuk0.gaijinsmash.riderz.ui.fragment.home

import android.app.Application
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.ViewModel
import android.arch.persistence.room.Room
import android.content.Context
import android.os.Bundle
import android.os.CountDownTimer
import android.util.Log
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide
import com.zuk0.gaijinsmash.riderz.R
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.Bsa
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.BsaXmlResponse
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Etd
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.EtdXmlResponse
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Trip
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.TripJsonResponse
import com.zuk0.gaijinsmash.riderz.data.local.room.database.FavoriteDatabase
import com.zuk0.gaijinsmash.riderz.data.remote.repository.BsaRepository
import com.zuk0.gaijinsmash.riderz.data.remote.repository.EtdRepository
import com.zuk0.gaijinsmash.riderz.data.remote.repository.TripRepository
import com.zuk0.gaijinsmash.riderz.ui.fragment.ParcelableViewModel
import com.zuk0.gaijinsmash.riderz.utils.SharedPreferencesUtils
import com.zuk0.gaijinsmash.riderz.utils.StationUtils
import com.zuk0.gaijinsmash.riderz.utils.TimeDateUtils
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/*
    ViewModel classes stores UI data and is lifecycle aware and can handle orientation changes.
    AndroidViewModel is for classes that need to use Context.
 */
@Singleton
class HomeViewModel @Inject
internal constructor(private val application: Application,
                     private val mTripRepository: TripRepository,
                     private val mBsaRepository: BsaRepository,
                     private val mEtdRepository: EtdRepository) : ViewModel(), ParcelableViewModel {

    var mBundle : Bundle = Bundle()
    var mBsaList : List<Bsa> = ArrayList()

    override fun writeTo(bundle: Bundle) {
        mBundle = bundle
    }

    override fun readFrom(bundle: Bundle) {

    }

    val bsaLiveData: LiveData<BsaXmlResponse>
        get() {
            return mBsaRepository.bsa
        }

    // For the home screen banner picture
    val hour: Int
        get() = TimeDateUtils.getCurrentHour()

    // Create message for Advisory Time and Date
    fun is24HrTimeOn(context: Context): Boolean {
        return SharedPreferencesUtils.getTimePreference(context)
    }

    private fun initTime(is24HrTimeOn: Boolean, time: String): String {
        val result: String

        if (is24HrTimeOn) {
            result = TimeDateUtils.format24hrTime(time)
        } else {
            result = TimeDateUtils.convertTo12Hr(time)
        }
        return result
    }

    fun initMessage(context: Context, is24HrTimeOn: Boolean, time: String): String {
        return context.resources.getString(R.string.last_update) + " " + initTime(is24HrTimeOn, time)
    }

    fun initPic(context: Context, hour: Int, imageView: ImageView) {
        if (hour < 6 || hour >= 21) {
            // show night picture
            Glide.with(context)
                    .load(R.drawable.sf_night)
                    .into(imageView)
        } else if (hour >= 17) {
            // show dusk picture
            Glide.with(context)
                    .load(R.drawable.sf_dusk)
                    .into(imageView)
        } else {
            Glide.with(context)
                    .load(R.drawable.sf_day)
                    .into(imageView)
        }
    }

    fun doesPriorityExist() : Boolean {
        val db = Room.databaseBuilder(application, FavoriteDatabase::class.java,
                "favorites").allowMainThreadQueries().build()
        val count = db.favoriteDAO.priorityCount
        db.close()
        if(count > 0) {
            return true
        }
        return false
    }

    fun getFavorite() : Favorite {
        val db = Room.databaseBuilder(application, FavoriteDatabase::class.java,
                "favorites").allowMainThreadQueries().build()
        val favorite = db.favoriteDAO.priorityFavorite
        db.close()
        return favorite
    }

    fun getEtdLiveData(origin: String): LiveData<EtdXmlResponse>? {
        val originAbbr = StationUtils.getAbbrFromStationName(origin)
        return mEtdRepository.getEtd(originAbbr)
    }

    fun createFavoriteInverse(trips: List<Trip>, favorite: Favorite) : Favorite {
        val trainHeaders = ArrayList<String>()
        for (trip in trips) {
            val header = trip.getLegList().get(0).getTrainHeadStation()
            if (!trainHeaders.contains(header)) {
                trainHeaders.add(header) // add a unique train header
                Log.i("HEADER", header)
            }
        }
        val favoriteInverse = Favorite()
        favoriteInverse.origin = favorite.getDestination()
        favoriteInverse.destination = favorite.getOrigin()
        favoriteInverse.trainHeaderStations = trainHeaders
        return favoriteInverse
    }

    fun getTripLiveData(origin: String, destination: String) : LiveData<TripJsonResponse>? {
        return mTripRepository.getTrip(StationUtils.getAbbrFromStationName(origin),
                StationUtils.getAbbrFromStationName(destination), "TODAY", "NOW")
    }

    fun getEstimatesFromEtd(favorite: Favorite, etdList: List<Etd>) : List<Estimate> {
        var results = ArrayList<Estimate>()
        for(etd in etdList) {
            if(favorite.trainHeaderStations.contains(etd.destinationAbbr)) {
                val estimate = etd.estimateList.get(0)
                estimate.origin = favorite.origin
                estimate.destination = favorite.destination
                results.add(estimate)
            }
        }
        return results
    }

    //todo: get current state of each timer
    //todo: if all timers are
    fun getRemainingTime() {

    }

    companion object {
        fun beginTimer(textView: TextView, minutesLeft: Int) {

            val untilFinished = (minutesLeft * 60000).toLong()
            object : CountDownTimer(untilFinished, 1000) {
                override fun onTick(millisUntilFinished: Long) {
                    val minutes = textView.context.resources.getString(R.string.minutes)
                    val seconds = textView.context.resources.getString(R.string.seconds)
                    textView.text = (millisUntilFinished / 60000).toString() + " " + minutes + " : " + millisUntilFinished % 60000 / 1000 + " " + seconds
                }

                override fun onFinish() {
                    textView.text = "Leaving!"
                    // todo: if leaving, wait 1 minute, then destroy itemListRow.
                    // must have reference to adapter.
                }
            }.start()
        }
    }

} // end of class
